sklearn/feature_extraction/dict_vectorizer.py

```diff
@@ -116,9 +121,18 @@ def fit(self, X, y=None):
             for f, v in six.iteritems(x):
                 if isinstance(v, six.string_types):
                     f = "%s%s%s" % (f, self.separator, v)
-                if f not in vocab:
-                    feature_names.append(f)
-                    vocab[f] = len(vocab)
+                if isinstance(v, Mapping):
+                    raise TypeError("DictVectorizer does not support objects as values.")
+                if hasattr(v, '__iter__') and not isinstance(v, Mapping):
```

Comment:
A numpy scalar (in v1.9.2, at least) has an attribute `'__iter__'`, i.e. `np.array(5).__iter__` succeeds. Checking that `iter` does not raise a `TypeError` would be correct.